src/libsync/bulkpropagatorjob.cpp

```diff
@@ -0,0 +1,869 @@
+            qCWarning(lcBulkPropagatorJob) << "Could not prepare upload device: " << device->errorString();
+
+            // If the file is currently locked, we want to retry the sync
+            // when it becomes available again.
+            if (FileSystem::isFileLocked(oneFile._localPath)) {
+                emit propagator()->seenLockedFile(oneFile._localPath);
+            }
+            // Soft error because this is likely caused by the user modifying his files while syncing
+            abortWithError(oneFile._item, SyncFileItem::SoftError, device->errorString());
+            return;
+        }
+        oneFile._headers["OC-Path"] = oneFile._remotePath.toUtf8();
+        uploadParametersData.push_back({std::move(device), oneFile._headers});
+    }
+
+    auto job = std::make_unique<PutMultiFileJob>(propagator()->account(), QUrl{}, std::move(uploadParametersData), this);
+    connect(job.get(), &PutMultiFileJob::finishedSignal, this, &BulkPropagatorJob::slotPutFinished);
+    connect(job.get(), &PutMultiFileJob::uploadProgress, this, &BulkPropagatorJob::slotUploadProgress);
+    //connect(job.get(), &PutMultiFileJob::uploadProgress, device.get(), &UploadDevice::slotJobUploadProgress);
+    connect(job.get(), &QObject::destroyed, this, &BulkPropagatorJob::slotJobDestroyed);
+    //adjustLastJobTimeout(job.get(), oneFile._fileSize);
+    auto jobCopy = job.get();
+    _jobs.append(job.release());
+    jobCopy->start();
+}
+
+void BulkPropagatorJob::slotComputeContentChecksum(SyncFileItemPtr item,
+                                                   UploadFileInfo fileToUpload)
+{
+    qDebug() << "Trying to compute the checksum of the file";
+    qDebug() << "Still trying to understand if this is the local file or the uploaded one";
+    if (propagator()->_abortRequested) {
+        return;
+    }
+
+    const QString filePath = propagator()->fullLocalPath(item->_file);
+
+    // remember the modtime before checksumming to be able to detect a file
+    // change during the checksum calculation - This goes inside of the item->_file
+    // and not the fileToUpload because we are checking the original file, not there
+    // probably temporary one.
+    item->_modtime = FileSystem::getModTime(filePath);
+
+    const QByteArray checksumType = propagator()->account()->capabilities().preferredUploadChecksumType();
+
+    // Maybe the discovery already computed the checksum?
+    // Should I compute the checksum of the original (item->_file)
+    // or the maybe-modified? (fileToUpload._file) ?
+
+    QByteArray existingChecksumType;
+    QByteArray existingChecksum;
+    parseChecksumHeader(item->_checksumHeader, &existingChecksumType, &existingChecksum);
+    if (existingChecksumType == checksumType) {
+        slotComputeTransmissionChecksum(item, fileToUpload, checksumType, existingChecksum);
+        return;
+    }
+
+    // Compute the content checksum.
+    auto computeChecksum = std::make_unique<ComputeChecksum>(this);
```

Comment:
I'm not sure I'm convinced by this pattern of using `std::unique_ptr` just to satisfy SonarCloud. ComputeChecksum's memory is already managed by the parent that got passed with this. I think doing both, could lead to really errors as well when you forget to release the pointer from the unique pointer.

Reply:
I am really convinced that using raw pointers is just an old habit and that we should really change and move out of this

Reply:
My last two cents on this ;)

Raw pointers are perfectly valid, to give for example others temporarily access to a resource. I think raw pointers are just disliked because they encourage manual memory management, but here we don't do manual memory management. Don't get me wrong, I love smart pointers, but Qt has its own memory management system and I think it's better to respect it, than mixing different memory management technologies. This can get very confusing. For example here you now have two owners of the memory that don't know from each other. Now you have to call `release()`, otherwise you have a bug. If you don't call it, you will have weird access violations that might be difficult to spot. On the other hand, the ComputeChecksum's memory is managed from the first moment on because it has a parent.

Reference:
https://isocpp.org/blog/2013/11/smart-pointers